lib/heidrun/mapping_tools/marc.rb

```diff
@@ -0,0 +1,204 @@
+      # with a particular name and 'tag' attribute.
+      #
+      # @example
+      #   The XML element for control field 007 has a 'name' of 'controlfield'
+      #   and a 'tag' of '007', so:
+      #     r.node.children
+      #      .select(&name_tag_condition('controlfield', '007')).first
+      #   gives you the matching element.
+      #
+      # @param name [String] The element name, without the namespace
+      # @param tag  [String] The value of the element's 'tag' attribute
+      def name_tag_condition(name, tag)
+        lambda { |node| node.name == name && node[:tag] == tag }
+      end
+
+      ##
+      # Return a lambda suitable for Array#select that gives the XML element
+      # with a particular name, with a 'tag' attribute matching the given regex
+      #
+      # @param name [String] The element name, without the namespace
+      # @param tag  [Regexp] The pattern for matching the tag
+      def name_tag_match_condition(name, pattern)
+        lambda { |node| node.name == name && node[:tag] =~ pattern }
+      end
+
+      ##
+      # Return a lambda suitable for Array#select, that gives the XML element
+      # for a datafield's subfield that has a particular code
+      #
+      # @param name [String] The element name, without the namespace
+      # @param code  [String] The value of the element's 'code' attribute
+      def subfield_code_condition(code)
+        lambda { |node| node.name == 'subfield' && node[:code] == code }
+      end
+
+      ##
+      # Return an Array of Element for the datafield with the given number
+      # (tag)
+      #
+      # @param r    [Krikri::XmlParser::Value] The record root element
+      # @param name [String] The XML element name, without the namespace
+      # @param tag  [String|Regexp] The value of the element's 'tag' attribute,
+      #                             or a Regexp to match it
+      # @return     [Array] of Element
+      def select_field(r, name, tag)
+        if tag.class == Regexp
+          select_cond = name_tag_match_condition(name, tag)
+        else
+          select_cond = name_tag_condition(name, tag)
+        end
+        r.node.children.select(&select_cond)
+      end
+
+      ##
+      # Return an Element for the datafield with the given number (tag)
+      #
+      # @param r   [Krikri::XmlParser::Value] The record root element
+      # @param tag [String|Regexp] The tag, e.g. '240' or /^78[07]$/
+      # @return    [Array] of Element, per .select_field
+      def datafield_els(r, tag)
+        select_field(r, 'datafield', tag)
+      end
+
+      ##
+      # Return the String value of the datafield with the given number (tag)
+      #
+      # An empty array is returned if the datafield does not exist.
+      #
+      # @param  r   [Krikri::XmlParser::Value] The record root element
+      # @param  tag [String] The tag, e.g. '001'
+      # @return     [Array] of String ([] if element does not exist)
+      def datafield_values(r, tag)
```

Comment:
This method is `#datafield_values`, while the following method is `#controlfield_value` - the pluralization should probably be consistent.